include/net/socket_offload_ops.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2018 Linaro Limited.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Socket Offload Redirect API
+ */
+
+#ifndef __SOCKET_OFFLOAD_OPS_H__
+#define __SOCKET_OFFLOAD_OPS_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <sys/types.h>
+#include <net/net_ip.h>
+
+/*
+ * It is assumed that these offload functions follow the
+ * POSIX socket API standard for arguments, return values and setting of errno.
+ */
+struct socket_offload {
```

Comment:
Is there a plan to have any offload-specific fields in this structure? Because at this point it seems to cover socket API, so perhaps we could call it `struct socket_api`?

I'm asking, because I have an image in my mind for improving current POSIX sockets/secure sockets coexitstence. Instead of having separate `#ifdef`s  in every socket API function, we could have wrappers like in `socket_offload.h` and separate set of of API structures for POSIX sockets, secure sockets and whatever offloaded sockets we want. Then we could simply switch between them during compilation (or even runtime in the future?).

Reply:
BTW, why is there no `poll`?

Reply:
> BTW, why is there no poll?

Well, SimpleLink offload only supports select(), and I'd rather have select() than poll(), as it seems to me that more socket based applications and tutorials expect select().
I agree that one or the other should be added (ideally both), and a "not implemented" error returned if the offload chip doesn't support one or the other (and, in future, could implement poll over select or vice-versa).

Reply:
Thanks for the feedback.

> Instead of having separate #ifdefs in every socket API function, we could have wrappers like in socket_offload.h and separate set of of API structures for POSIX sockets, secure sockets and whatever offloaded sockets we want. Then we could simply switch between them during compilation (or even runtime in the future?).

A good thought.   
This PR adds yet another #ifdef to socket.h, which is starting to get messy.

My original thought was that Zephyr zsock could be the default "socket provider", and both variants of (zsock, offload) can be registered via something like `socket_register(const struct socket_ops *ops)`.

Some capabilities to consider:
1) The SimpleLink offload driver handles secure sockets through the same set of APIs as non-secure (just different flags);
2) It is possible for a non-secure socket to be "upgraded" to a secure socket via a setsockopt() call. 
3) It is possible to mix secure and non secure client threads together in the same application.

With the global #ifdef CONFIG_NET_SOCKETS_SOCKOPT_TLS , type compile switches, I guess points 2) and 3) aren't yet possible.    Maybe a runtime switch could help.

If you are close to maybe posting a WIP PR with this `struct socket_api` concept, perhaps we can merge the two?

Reply:
> I agree that one or the other should be added (ideally both)

I think this is the right way, have both specified in the API, and leave up to implementation, which ones are implemented. I just wanted to point it out, that existing socket samples in Zephyr uses `poll`, as it is currently available there, so if we wanted to use samples out-of-the-box (or with minor modifications) with offloaded driver, we'd have to have poll implemented.

> Some capabilities to consider:
> The SimpleLink offload driver handles secure sockets through the same set of APIs as non-secure (just different flags);
> It is possible to mix secure and non secure client threads together in the same application.

That's the same (or similar) as we have in Zephyr now. Secure sockets API proposed recently can handle both, secure protocols and non secure. So you can use the same API to create regular TCP/UDP sockets and secure TLS/DTLS sockets in the same application. The decision which one should be created is made based on protocol type provided during socket creation. In case TCP/UDP is used, socket API calls are simply forwarded to the `zsock` implementation. There is no example showing this functionality yet, but it is possible.

> It is possible for a non-secure socket to be "upgraded" to a secure socket via a setsockopt() call.

That's indeed impossible with current implmentation.

> With the global #ifdef CONFIG_NET_SOCKETS_SOCKOPT_TLS , type compile switches, I guess points 2) and 3) aren't yet possible. Maybe a runtime switch could help.

I hope I've already answered this. Let me know is something is still unclear.

> If you are close to maybe posting a WIP PR with this struct socket_api concept, perhaps we can merge the two?

Well, it was just general idea and I dont' have any code yet, and I'm afraid I can't do it now. I'm going on a vacation in a week and I have other work to finish before that. But as I said, I view this PR as a good base for future improvement, and this can be one of them :)